common/src/main/java/com/wynntils/features/user/overlays/GameNotificationOverlayFeature.java

```diff
@@ -0,0 +1,175 @@
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.ListIterator;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+
+@FeatureInfo(category = "Overlays")
+public class GameNotificationOverlayFeature extends UserFeature {
+    private static GameNotificationOverlayFeature INSTANCE;
+    private static final List<MessageContainer> messageQueue = new LinkedList<>();
+
+    @OverlayInfo(renderType = RenderEvent.ElementType.GUI)
+    public final GameNotificationOverlay gameNotificationOverlay = new GameNotificationOverlay();
+
+    @SubscribeEvent
+    public void onGameNotification(NotificationEvent.Queue event) {
+        messageQueue.add(event.getMessageContainer());
+
+        if (GameNotificationOverlayFeature.INSTANCE.gameNotificationOverlay.overrideNewMessages
+                && messageQueue.size() > GameNotificationOverlayFeature.INSTANCE.gameNotificationOverlay.messageLimit) {
+            messageQueue.remove(0);
+        }
+    }
+
+    @SubscribeEvent
+    public void onGameNotification(NotificationEvent.Edit event) {
+        MessageContainer newContainer = event.getMessageContainer();
+        messageQueue.stream()
+                .filter(messageContainer -> messageContainer.hashCode() == newContainer.hashCode())
+                .findFirst()
+                .ifPresent(messageContainer -> {
+                    messageContainer.update(newContainer);
+                });
+    }
+
+    public static class GameNotificationOverlay extends Overlay {
+
+        @Config
+        public int messageLimit = 5;
+
+        @Config
+        public boolean invertGrowth = true;
+
+        @Config
+        public int messageMaxLength = 0;
+
+        @Config
+        public FontRenderer.TextShadow textShadow = FontRenderer.TextShadow.OUTLINE;
+
+        @Config
+        public boolean overrideNewMessages = true;
+
+        private TextRenderSetting textRenderSetting;
+
+        public GameNotificationOverlay() {
+            super(
+                    new OverlayPosition(
+                            0,
+                            0,
+                            VerticalAlignment.Top,
+                            HorizontalAlignment.Right,
+                            OverlayPosition.AnchorSection.BottomRight),
+                    new GuiScaledOverlaySize(250, 110));
+
+            updateTextRenderSetting();
+        }
+
+        @Override
+        public void render(PoseStack poseStack, float partialTicks, Window window) {
+            List<MessageContainer> toRender = new ArrayList<>();
+
+            ListIterator<MessageContainer> messages = messageQueue.listIterator(messageQueue.size());
+            while (messages.hasPrevious()) {
+                MessageContainer message = messages.previous();
+
+                if (message.getRemainingTime() <= 0.0f) {
+                    messages.remove(); // remove the message if the time has come
```

Comment:
Something does still not feel right about the design. It should not be the responsibility of the Overlay to remove messages, if the timer is in the messages themselves. 

There is something schizophrenic about the notifications, how they are partially a "core" functionality, but still tightly coupled to how they are displayed. I think either the core notification system should only accept new messages and pass them on, without knowing anything about timers, or the queue of not-yet-expired notifications should reside in the notification manager. (I think I like the latter approach better).

Reply:
I agree with removing message queue in the NotificationManager and only sending out events. But I think features consuming these messages can have their own system of storing them.

Reply:
I'm still not convinced but I'm prepared to turn a blind eye for this time only. ;-)

Reply:
Just to clarify: I could be okay with the Feature keeping a queue of notifications, and the manager just responsible for sending them out as they arrive.

But, then the time delay should be kept out of the manager, since that is part of the overlay queue handling.

Reply:
That is the case?

Reply:
Oh well there is a time limit there too. I can remove that I guess.